File: src/CSharp/594.LongestHarmoniousSubsequence.cs


using System;
using System.Linq;
using System.Collections.Generic;

namespace PeefyLeetCode.LongestHarmoniousSubsequence
{
    public class Solution
    {
        public int FindLHS(int[] nums)
        {
            if (nums == null || nums.Length == 0 || nums.Length == 1)
            {
                return 0;
            }
            Array.Sort(nums);
            int max = 0, temp = nums[0], begin = 0, index = 0;
            bool isStart = true;
            for (int i = 1; i <= nums.Length; i++)
            {
                if (i == nums.Length || nums[i] != temp)
                {
                    if (!isStart)
                    {
                        max = max > (i - begin) ? max : (i - begin);
                        begin = index;
                    }
                    if (i == nums.Length) break;
                    if (nums[i] - temp == 1)
                    {
                        isStart = false;
                    }
                    else
                    {
                        isStart = true;
                        begin = i;
                    }
                    temp = nums[i];
                    index = i;
                }
            }
            return max;
        }
    }
}

File: src/Python/594.LongestHarmoniousSubsequence.py

from collections import Counter

class Solution:
    def findLHS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        count = 0
        d = Counter(nums)
        for v in d.keys():
            if v - 1 in d.keys() and d[v] + d[v - 1] > count:
                count = d[v] + d[v - 1]
        return count

if __name__ == '__main__':
    solution = Solution()
    print(solution.findLHS([1,3,2,2,5,2,3,7]))
else:
    pass
